plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/ModelExecutorThread.java

```diff
@@ -76,24 +79,20 @@ private void newMailboxEntry(MailboxEntry entry) {
 	 * becomes empty.
 	 */
 	void shutdown() {
-		try {
-			class ShutdownAction implements MailboxEntry {
-				@Override
-				public void execute() {
-					if (mailbox.isEmpty()) {
-						ModelExecutor.shutdownNow();
-					} else {
-						try {
-							mailbox.put(new ShutdownAction());
-						} catch (InterruptedException e) {
-						}
-					}
-				}
-			}
+		shutdownWhenEmpty = true;
+		// wake up if waiting with an empty message
+		mailbox.add(() -> {});
+	}
 
-			mailbox.put(new ShutdownAction());
-		} catch (InterruptedException e) {
-		}
+	/**
+	 * Sets the model executor to be shut down immediately. If the execution of
+	 * a {@link #mailbox} item does not terminate it cannot guarantee the
+	 * termination.
+	 */
+	void shutdownImmediately() {
+		shutdownImmediately = true;

```

Comment:
This method can be called from other threads. Changing a field without any synchronization is unsafe.